### ISSUE TITLE
ChatService().getChatByStream(symMessage.getStreamId()) intermittently null

### ISSUE THREAD

**Issue by peterjoel** [symphonyoss/symphony-java-client#33]
See:
https://gist.github.com/peterjoel/e7a2e53017bd3a1d720b9b163868f6a7

Inside onChatMessage(), chat is sometimes null but sometimes it is a valid Chat object.

**Comment by ftbb** [comment]
Thanks for providing example code.  

You get to look at my new wiki page on how the java-client works.  Please look at second diagram on [overview page](https://github.com/symphonyoss/symphony-java-client/wiki/Overview)

The issue is that you are registering a listener on the MessageService vs ChatService:

    symphonyClient.getMessageService().addChatListener(this);

The ChatService is also listening to the MessageService, which means both your application and the ChatService are receiving the message event at about the same time.  The ChatService has yet to register the Chat object (model) for that stream from the incoming message, which is an issue when you are immediately querying for this chat object from the ChatService.  

Note: The ChatService will construct a full Chat objects from the incoming message

Actions:
1. Please add the ChatServiceListener to the application. Leave the ChatListener as well.

2. Change symphonyClient.getMessageService().addChatListener(this); to symphonyClient.getChatService.addListener(this); 

3. Within the implemented callbacks:
      
       void onNewChat(Chat chat){ 
         LOG.info("onNewChat: {} {}", chat.getStreamId(), chat.getLastMessage());
        
        //This should never report null
        Chat chat = symphonyClient.getChatService().getChatByStream(symMessage.getStreamId());
        if(chat == null){
            LOG.error("CHAT IS NULL");
        }
        else {
            LOG.info("Chat is non-null");
        }
                 //Register the ChatListener to the individual chat object
                 chat.addListener(this); 
        } 
        
       //You have to implement this to receive message from Chat objects you have registered the listener to
       void onChatMessage(SymMessage message){
                
          LOG.info("Message received from: {} message {}",message.getStreamId(), message.getMessage());
          
           //Lets pull the chat object for this stream...
          Chat chat = symphonyClient.getChatService().getChatByStream(symMessage.getStreamId());
      
       }
  
In short registering addChatListener(...) in MessageService will listen to all raw incoming chat messages (sans room).  The ChatService is at the same level, which means there is no guarantee that actual Chat object would be created at that time. 

Please test and verify...  If you are happy with the response, please close ticket.

**Comment by peterjoel** [comment]
Thanks. The API is a little confusing here. But it works now, like you say.

**Comment by ftbb** [comment]
Your feedback is great, so we will look to making it more obvious in future releases.

**Comment by peterjoel** [comment]
@ftbb will you accept PRs for API changes? Do you have a chat room, or other place, for these types of conversation?

**Comment by ftbb** [comment]
Yes, we accept PR if you have an ICLA or CCLA signed with the foundation.

Best way to reach me is over Symphony directly.
